In [1]:
import pandas

rating = pandas.read_table("ratings.csv", sep=",")

rating.iloc[:5]

UserID  MovieID  Rating  Timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [2]:
rating.tail()

UserID  MovieID  Rating   Timestamp
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

In [3]:
rating.isnull().sum().sum()

0

In [4]:
total_user = rating["UserID"].nunique()

total_user

610

In [5]:
total_film = rating["MovieID"].nunique()

total_film

9724

In [6]:
rata_rating = rating["Rating"].mean().round(3)

rata_rating

3.502

On Average, Our Users Have Given a Rating of 3.5 out of 5.0

In [7]:
rating_table = rating.pivot_table(index="MovieID", columns="UserID", values="Rating")

SparseRate = lambda table : round((table.isnull().sum().sum() / table.size) * 100, 3)

SparseRate(rating_table)

98.3

Our 98.3 % Sparsity Rate Signaling Many Things.

* Our Table Mostly Consists of 0

* Our Films Haven't been Watched

* Our Films is Watched, but the User Didn't Rate the Film

* One or Two Films is Watch Less than 10 Times or User Only Watch Less Than 50 Movie

Our Solution is to Set a Minimum Threshold to Film - User Interaction

The First Threshold is Film. Our Minimum Threshold for a Film is

*This Film Must be Watched, Minimum of 10 Times*

In [8]:
minimum_film_interaction = 10

WatchFilm = rating.groupby("MovieID")["Rating"].agg(["count", "mean"])

WatchFilm.columns = ["TotalWatch", "Rating"]

WatchFilm = WatchFilm[WatchFilm["TotalWatch"] > minimum_film_interaction]

rating = rating[rating["MovieID"].isin(WatchFilm.index)]

rating["MovieID"].nunique()

2121

The Last Threshold is User. Our Minimum Threshold for a User is

*This User Must be Watch & Rate, Minimum of 50 Films*

In [9]:
minimum_user_interaction = 50

UserWatch = rating.groupby("UserID")[["Rating"]].count()

UserWatch = UserWatch[UserWatch["Rating"] > minimum_user_interaction]

rating = rating[rating["UserID"].isin(UserWatch.index)]

In [10]:
rating_table = rating.pivot_table(index="MovieID", columns="UserID", values="Rating")

SparseRate(rating_table)

90.396

Our Final Sparsity Rate is 90.39 %

In [11]:
rating_table.iloc[:5, :8]

UserID    1   4    6    7   10   11   15  16
MovieID                                     
1        4.0 NaN  NaN  4.5 NaN  NaN  2.5 NaN
2        NaN NaN  4.0  NaN NaN  NaN  NaN NaN
3        4.0 NaN  5.0  NaN NaN  NaN  NaN NaN
5        NaN NaN  5.0  NaN NaN  NaN  NaN NaN
6        4.0 NaN  4.0  NaN NaN  5.0  NaN NaN

In [12]:
rating_table = rating_table.fillna(0)

rating_table.iloc[:5, :8]

UserID    1    4    6    7    10   11   15   16
MovieID                                        
1        4.0  0.0  0.0  4.5  0.0  0.0  2.5  0.0
2        0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0
3        4.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0
5        0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0
6        4.0  0.0  4.0  0.0  0.0  5.0  0.0  0.0

In [13]:
rating_table.index.name

'MovieID'

In [14]:
rating_table.columns.name

'UserID'

In [15]:
usecols = ["MovieID", "Title"]

table_film = pandas.read_table("movies.csv", sep=",", usecols=usecols)

rating = rating.join(table_film.set_index("MovieID"), on="MovieID")

rating.iloc[:5]

UserID  MovieID  Rating  Timestamp                        Title
0       1        1     4.0  964982703             Toy Story (1995)
1       1        3     4.0  964981247      Grumpier Old Men (1995)
2       1        6     4.0  964982224                  Heat (1995)
3       1       47     5.0  964983815  Seven (a.k.a. Se7en) (1995)
4       1       50     5.0  964982931   Usual Suspects, The (1995)

What is Sparse Matrix ? Sparse Matrix is Matrix that Many of the Value is 0

Usually Sparse Matrix is Mostly Use in Collaborative Method Recommendation System

In [16]:
from scipy.sparse import csr_matrix

rating_matrix = csr_matrix(rating_table.values)

rating_matrix.shape

(2121, 352)

In [17]:
rating_reset = rating_table.reset_index()

rating_reset.shape

(2121, 353)

In [18]:
rating_reset.iloc[:5, :10]

UserID  MovieID    1    4    6    7   10   11   15   16   17
0             1  4.0  0.0  0.0  4.5  0.0  0.0  2.5  0.0  4.5
1             2  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0
2             3  4.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0
3             5  0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0
4             6  4.0  0.0  4.0  0.0  0.0  5.0  0.0  0.0  0.0

In [19]:
rating_reset.columns.name

'UserID'

In [20]:
rating_table.iloc[:10]

UserID   1    4    6    7    10   11   15   16   17   18   ...  600  601  602  \
MovieID                                                    ...                  
1        4.0  0.0  0.0  4.5  0.0  0.0  2.5  0.0  4.5  3.5  ...  2.5  4.0  0.0   
2        0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  3.0  ...  4.0  0.0  4.0   
3        4.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5        0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  2.5  0.0  0.0   
6        4.0  0.0  4.0  0.0  0.0  5.0  0.0  0.0  0.0  4.0  ...  0.0  0.0  3.0   
7        0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  3.5  0.0  0.0   
9        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
10       0.0  0.0  3.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  3.0   
11       0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  3.0   
12       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

UserID   603  604  605  606  607  608  610  
MovieID                                     
1        4.0  3.0  4.0  2.5  4.0  2.5  5.0  
2        0.0  5.0  3.5  0.0  0.0  2.0  0.0  
3        0.0  0.0  0.0  0.0  0.0  2.0  0.0  
5        0.0  3.0  0.0  0.0  0.0  0.0  0.0  
6        4.0  3.0  0.0  0.0  0.0  0.0  5.0  
7        0.0  0.0  0.0  2.5  0.0  0.0  0.0  
9        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
10       0.0  0.0  0.0  0.0  0.0  4.0  0.0  
11       0.0  0.0  0.0  2.5  3.0  0.0  0.0  
12       0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[10 rows x 352 columns]

In [21]:
rating_reset.iloc[:10]

UserID  MovieID    1    4    6    7   10   11   15   16   17  ...  600  601  \
0             1  4.0  0.0  0.0  4.5  0.0  0.0  2.5  0.0  4.5  ...  2.5  4.0   
1             2  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  4.0  0.0   
2             3  4.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
3             5  0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  2.5  0.0   
4             6  4.0  0.0  4.0  0.0  0.0  5.0  0.0  0.0  0.0  ...  0.0  0.0   
5             7  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  3.5  0.0   
6             9  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
7            10  0.0  0.0  3.0  0.0  0.0  3.0  0.0  0.0  0.0  ...  0.0  0.0   
8            11  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
9            12  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

UserID  602  603  604  605  606  607  608  610  
0       0.0  4.0  3.0  4.0  2.5  4.0  2.5  5.0  
1       4.0  0.0  5.0  3.5  0.0  0.0  2.0  0.0  
2       0.0  0.0  0.0  0.0  0.0  0.0  2.0  0.0  
3       0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  
4       3.0  4.0  3.0  0.0  0.0  0.0  0.0  5.0  
5       0.0  0.0  0.0  0.0  2.5  0.0  0.0  0.0  
6       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
7       3.0  0.0  0.0  0.0  0.0  0.0  4.0  0.0  
8       3.0  0.0  0.0  0.0  2.5  3.0  0.0  0.0  
9       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[10 rows x 353 columns]